### ISSUE TITLE
proposal: `is_check_on_generic_argument` - flag `T is Foo` when `T` is a type variable

### ISSUE THREAD

**Issue by natebosch** [dart-lang/linter#4480]
```dart
class C<T> {
  void something() {
    if (T is Foo) {} // LINT
    if (T is! Foo) {} // LINT
  }
}
```

I don't think there are any useful is checks on generic type variables. `T is Type` and `T is! AnythingElse` will always be `true`.

It is possible to do `T == AnythingElse`, but this is also risky because it does not handle subtypes like `is` would.

**Comment by lrhn** [comment]
Could this be special case of a more general "this test is always false" lint?

We know that type literals have exact types (not really, but there are no *public* types that it implements), so `T is Something` is wrong for all non-`Type` types.
There are other tests that we can *probably* say are never true (related to `final`/`sealed` types that cannot be subtyped outside of the declaring library).

Even if not, doing `TypeVar is Type` is a common enough mistake that maybe it should be an always-on analyze warning, not a lint.

**Comment by bwilkerson** [comment]
> Could this be special case of a more general "this test is always false" lint?

It would be easier to answer that question if we had a reasonably complete list of the other cases that the lint would catch, but I think the answer is probably 'no'. Users probably want more fine grained control over what's reported and what isn't than a single monolithic rule would provide. But asking whether there are closely enough related issues that a lint like this could catch is useful, because users also don't want to have to enable multiple lints for what they perceive to be a single kind of problem.

**Comment by natebosch** [comment]
> doing `TypeVar is Type` is a common enough mistake that maybe it should be an always-on analyze warning, not a lint.

I found 7 files using this pattern internally. I haven't checked any external corpus yet.

@bwilkerson - WDYT about including this narrow case as an always-on diagnostic?

**Comment by bwilkerson** [comment]
Given that there are (as I understand it) no false positives, I think it makes sense for it to be a warning.

**Comment by natebosch** [comment]
Cl open to add as a warning at https://dart-review.googlesource.com/c/sdk/+/311924

We will want to clean up some existing cases before landing it I think.
